docs/proposals/parallel-compaction.md

```diff
@@ -0,0 +1,68 @@
+
+### Problem and Requirements
+Currently, a compactor is able to compact up to 20M timeseries within 2 hours for a level-2 compaction, including the time to download blocks, compact, and upload the newly compacted block. We would like to increase the timeseries limit per tenant, and compaction is one of the limiting factors. In addition, we would like to achieve the following:
+
+* Compact multiple non-overlapping time intervals concurrently, so we can achieve higher throughput for the compaction of a single tenant
+* We should be able to scale up, down compactor as needed, depending on how many compactions are pending
+* Insight into the compaction progress of a tenant, such as the number of compactions required in order to catch up to the newest blocks
+
+## Design
+
+We accept the fact that a single compaction can potentially take more than 2 hours to compact, and we achieve higher compaction throughput through horizontally scaling the compactor. To compact more blocks in parallel for a single tenant, we distribute the compaction groups to compactors, instead of introducing more parallelism within a compactor.
+
+### Parallelize Work
+
+This proposal builds heavily on top of the [GrafanaLabs approach of introducing parallelism via time intervals](https://github.com/cortexproject/cortex/pull/2616). The difference being that a single tenant is now sharded across multiple compactors instead of just a single compactor. The initial approach will be to work on distinct time intervals, but the compactor planner can be later extended to introduce parallelism within a time interval as well. The
+
+The following is an example of parallelize work at each level:
+
+![Parallel Compaction Grouping](/images/proposals/parallel-compaction-grouping.png)
+
+Compactors are shuffle-sharded, meaning that 1 tenant can belong to multiple compactors, and these subset of compactors determine which blocks should be compacted together. Compactors determine amongst themselves the responsibility of the compaction blocks, by using a hash on time interval and the block ids.
+
+The benefit of this approach is that this aligns with what Cortex currently does in Ruler. The downside is that a compaction job can only be assigned to a single compactor, rather than all of the compactors sharded for the tenant. If a compaction job takes forever, other tenants sharded to the same compactor will be blocked until the issue is resolved. With the scheduler approach, any compactor assigned to a given tenant can pick up any work required.
```

Comment:
What you describe can be soften adding both horizontal sharding and parallelisation in the compactor. If you have 4 parallel workers in the compactor, then even if 1 is blocked for a very long time, other tenants blocks are still compacted by the remaining 3 workers.

You could obviously have all 4 workers blocked but chances are lower.

Reply:
I'm afraid of introducing too much parallelism within a compactor. It might making capacity management for compactor harder, since we can potentially have multiple large users in the same compactor at the same time, and causing OOM or out-of-disk-space issue for each other.

If we run into this issue, I think having the scheduler coordinating the compactors will be easier for capacity management, and solving the issue once and for all. Especially we can opt for smaller but more compactors, instead of larger compactors.